src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -31,11 +31,112 @@ public struct CallGraphNode
+        /// simultaneously treated as concrete types. 
+        /// NOTE: This routine prioritizes the verifications to ensure the correctness of the resolution over performance.  
+        /// </summary>
+        public static bool TryCombineTypeResolutions
+            (QsQualifiedName parent, out ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType> combined,
+            params ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>[] resolutions)
+        {
+            if (parent == null) throw new ArgumentNullException(nameof(parent));
+            var combinedBuilder = ImmutableDictionary.CreateBuilder<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>();
+            var success = true;
+
+            static Tuple<QsQualifiedName, NonNullable<string>> AsTypeResolutionKey(QsTypeParameter tp) => Tuple.Create(tp.Origin, tp.TypeName);
+            static bool ResolutionToTypeParameter(Tuple<QsQualifiedName, NonNullable<string>> typeParam, ResolvedType res) =>
+                res.Resolution is ResolvedTypeKind.TypeParameter tp && tp.Item.Origin.Equals(typeParam.Item1) && tp.Item.TypeName.Equals(typeParam.Item2);
+
+            // Returns true if the given resolution for the given key constrains the type parameter 
+            // by mapping it to a different type parameter belonging to the same callable. 
+            bool InconsistentResolutionToNative(Tuple<QsQualifiedName, NonNullable<string>> key, ResolvedType resolution)
+            {
+                var resolutionToTypeParam = resolution.Resolution as ResolvedTypeKind.TypeParameter;
+                var isResolutionToNative = resolutionToTypeParam != null && resolutionToTypeParam.Item.Origin.Equals(parent);
+                return isResolutionToNative
+                    // We can omit this check as long as combinedBuilder only ever contains native type parameters:
+                    // && key.Item1.Equals(parent) 
+                    && key.Item2.Value != resolutionToTypeParam.Item.TypeName.Value;
+            }
+
+            foreach (var resolution in resolutions)
+            {
+                // Contains a lookup of all the keys in the combined resolutions whose value needs to be updated 
+                // if a certain type parameter is resolved by the currently processed dictionary. 
+                var mayBeReplaced = combinedBuilder
+                    .Where(kv => kv.Value.Resolution.IsTypeParameter)
+                    .ToLookup(
+                        kv => AsTypeResolutionKey(((ResolvedTypeKind.TypeParameter)kv.Value.Resolution).Item),
+                        entry => entry.Key);
+
+                // We need to ensure that the mappings for external type parameters are processed first, 
+                // to cover an edge case that would otherwise be indicated as a conflicting resolution.
```

Comment:
I'm still not getting this. Why do you need to process this for loop first, before the other one? To my thinking you would need to process the other for loop first. Perhaps an explanation of this edge case mentioned here would help clear this up?

Reply:
If you don't sort or change which loop is executed first, you will see which test fails. :)